f3/granite.go

```diff
@@ -350,8 +369,101 @@ func (i *instance) isValid(msg *GMessage) bool {
 	return true
 }
 
-// TODO Checks whether a message is justified by prior messages.
+// Checks whether a message is justified.
 func (i *instance) isJustified(msg *GMessage) bool {
+
```

Comment:
My initial reaction was that this is hard to read. I will elaborate.

Reply:
My suggestion would be to use errors instead of logs (no error = justified).

Reply:
Done, but I swapped your suggestion: because some conditions of semantic validation do not require (or have) any actual aggregate to verify, I instead have the `isJustified` perform semantic validation and call a new function `verifyJustification` at the end (if it has not returned with nil or error already before).